caffe2/operators/mean_op.h

```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
 
   template <typename T>
   bool DoRunWithType() {
-    auto& dY = Input(0);
-    const auto* dY_data = dY.template data<T>();
-    int size = dY.numel();
-
     int num_inputs = OutputSize();
     float scale = 1.0f / num_inputs;
-
-    // dX0 = scale * dY
-
-    auto* dX0 = Output(0, dY.sizes(), at::dtype<T>());
-    math::Scale(
-        size, scale, dY_data, dX0->template mutable_data<T>(), &context_);
-
-    // Copy the rest dX
-    for (int i = 1; i < num_inputs; i++) {
-      auto* cur_dX = Output(i);
-      cur_dX->ResizeLike(dY);
-      cur_dX->CopyFrom(*dX0, true /*async*/);
+    std::vector<int> output_dims;
+    auto& dY = Input(0);
+    const auto* dY_data = dY.template data<T>();
+    std::copy(
+        dY.sizes().cbegin(),
+        dY.sizes().cend(),
+        std::back_inserter(output_dims));
+    for (int i = num_inputs; i >= 1; i--) {
```

Comment:
in case Mean(a, b, c), GI(1) (i.e., GI(b)) should be computed as GO(0) / input_num. If the shape doesn't match, (broadcasting happens), you should use reduce sum to get the final gradient for b. the same applies to a, c.

Reply:
I think this is what the code is doing. Note that `dY = Input(0) = GO(0)`, and I use reduce sum in line 148 with input `dY` and the current input `dX = GI(i-1)` broadcasting it to dimensions specified by `Input(i) = I(i-1)` . One thing I'm not doing is checking if the dimensions are equal, to avoid the ReduceSum computation when it is unnecessary. 

Also note that the `op_mean_test.py` has a gradient check that includes a size check, and the test passes. @BIT-silence is there something missing on those tests?